code/modules/power/apc/apc_main.dm

```diff
@@ -0,0 +1,567 @@
+			equipment = autoset(equipment, AUTOSET_OFF)
+			lighting = autoset(lighting, AUTOSET_OFF)
+			environ = autoset(environ, AUTOSET_ON)
+			alarm_manager.send_alarm(ALARM_POWER)
+		else if(cell.percent() < 30 && long_term_power < 0) // <30%, turn off equipment
+			equipment = autoset(equipment, AUTOSET_OFF)
+			lighting = autoset(lighting, AUTOSET_ON)
+			environ = autoset(environ, AUTOSET_ON)
+			alarm_manager.send_alarm(ALARM_POWER)
+		else // otherwise all can be on
+			equipment = autoset(equipment, AUTOSET_ON)
+			lighting = autoset(lighting, AUTOSET_ON)
+			environ = autoset(environ, AUTOSET_ON)
+			if(cell.percent() > 75)
+				alarm_manager.clear_alarm(ALARM_POWER)
+
+
+		// now trickle-charge the cell
+		if(chargemode && charging == APC_CHARGING && operating)
+			if(excess > 0) // check to make sure we have enough to charge
+				// Max charge is capped to % per second constant
+				var/ch = min(excess JOULES, cell.maxcharge JOULES)
+				add_load(ch WATTS) // Removes the power we're taking from the grid
+				cell.give(ch) // actually recharge the cell
+
+			else
+				charging = APC_NOT_CHARGING // stop charging
+				chargecount = 0
+
+		// show cell as fully charged if so
+		if(cell.charge >= cell.maxcharge)
+			cell.charge = cell.maxcharge
+			charging = APC_FULLY_CHARGED
+
+		if(chargemode)
+			if(!charging)
+				if(excess > cell.maxcharge*GLOB.CHARGELEVEL)
+					chargecount++
+				else
+					chargecount = 0
+
+				if(chargecount == 10)
+
+					chargecount = 0
+					charging = APC_CHARGING
+
+		else // chargemode off
+			charging = APC_NOT_CHARGING
+			chargecount = 0
+
+	else // no cell, switch everything off
+
+		charging = APC_NOT_CHARGING
+		chargecount = 0
+		equipment = autoset(equipment, AUTOSET_FORCE_OFF)
+		lighting = autoset(lighting, AUTOSET_FORCE_OFF)
+		environ = autoset(environ, AUTOSET_FORCE_OFF)
+		alarm_manager.send_alarm(ALARM_POWER)
+
+	// update icon & area power if anything changed
+
+	if(last_lt != lighting || last_eq != equipment || last_en != environ || force_update)
+		force_update = 0
+		queue_icon_update()
+		update()
+	else if(last_ch != charging)
+		queue_icon_update()
+
+/obj/machinery/power/apc/proc/reset(wire)
+	switch(wire)
+		if(WIRE_IDSCAN)
+			locked = TRUE
+		if(WIRE_POWER1, WIRE_POWER2)
+			if(!wires.is_cut(WIRE_POWER1) && !wires.is_cut(WIRE_POWER2))
+				shorted = FALSE
+		if(WIRE_AI)
+			if(!wires.is_cut(WIRE_AI))
+				aidisabled = FALSE
+		if(APC_RESET_EMP)
+			equipment = APC_CHANNEL_AUTO_ON
+			environ = APC_CHANNEL_AUTO_ON
+			update_appearance()
+			update()
+
+// overload all the lights in this APC area
+/obj/machinery/power/apc/proc/overload_lighting()
+	if(!operating || shorted)
+		return
+	if(cell && cell.charge >= 20)
+		cell.use(20)
+		INVOKE_ASYNC(src, .proc/break_lights)
+
+/obj/machinery/power/apc/proc/break_lights()
+	for(var/obj/machinery/light/breaked_light as anything in area)
+		breaked_light.on = TRUE
+		breaked_light.break_light_tube()
+		stoplag()
```

Comment:
Just a heads up, the `as anything in area` here is causing runtimes. (On a downstream technically, but the code is untouched.)
![image](https://user-images.githubusercontent.com/57483089/163883304-6c7a3054-3d36-40a4-bea0-6838d3c1d7cd.png)